src/middleware/authMiddleware.js

```diff
@@ -1,35 +1,92 @@
 const jwt = require("jsonwebtoken");
-
+const { getPostById } = require("../services/postService");
+const { getUserByUsername } = require("../services/userService");
 
 const authenticate = (req, res, next) => {
-    const token = req.headers?.authorization && req.headers.authorization.split(" ")[1];
+    const token = getToken(req);
     if (!token) {
         return res.status(401).json("Unauthorized Access");
     }
+
     try {
         const user = jwt.verify(token, process.env.JWT_SECRET);
         res.locals.user = user;
         next();
     } catch (err) {
         return res.status(401).json("Unauthorized Access, try relogging");
     }
-}
+};
+
+const accountOwnerAuthenticate = async (req, res, next) => {
+    const token = getToken(req);
+    if (!token) {
+        return res.status(401).json("Unauthorized Access");
+    }
+
+    try {
+        const user = jwt.verify(token, process.env.JWT_SECRET);
+
+        const userId = req.params.userId;
+        if (userId !== user.itemID) {
+            return res.status(401).json("Unauthorized Access - Wrong User");
+        }
+
+        res.locals.user = user;
+        next();
+    } catch (err) {
+        return res.status(401).json("Unauthorized Access, try relogging");
+    }
+};
+
+const postOwnerAuthenticate = async (req, res, next) => {
```

Comment:
I do not think we need these additional auth functions unless there's a unique use case for why you did this. You can achieve the same result with default auth.

Reply:
The default auth doesn't check if the sender of the request is the owner of an account/post, so that would have to be checked in the controller or the service, when it should not be the job of those layers to check for authorization.

Reply:
I made a mistake when I created the adminAuth middleware. All these functions essentially call Authenticate and then have a single if check. 
To reduce the code copy paste, maybe we keep the regular auth function and move the if checks to their own middleware file such as permissionsMiddleware or authorizationMiddleware. That way we can just have the if statements isolated. So for adminAuthenticate, it be something like this

``` Javascript
function roleCheck(req, res, next) {
    if (res.locals.user.role !== "admin") {
        return res.status(401).json({message: "unauthorized"});
    }
    next()
}
```

We would have to ensure we use the authenticate middleware before these

Reply:
Oooh and with the ones that need post/user info once you get the info you can put in res.locals and we won't have to call the database again in the controller.

``` Javascript
function postOwnerCheck(req, res, next) {
    try {
        const post = await getPostById(postId);
        res.locals.post = post; // res.locals.otherUser = otherUser if needed
        userId = (await getUserByUsername(post.postedBy)).itemID;
    } catch (err) {
        return res.status(err.status).json({ message: err.message });
    }
}
```